plugin.js

```diff
@@ -27,15 +29,33 @@ function factory (options) {
     }
 
     const key = header.substring(bearerType.length).trim()
-    if (authenticate(keys, key) === undefined) {
-      const invalidKeyError = Error('invalid authorization header')
-      fastifyReq.log.error('invalid authorization header: `%s`', header)
-      if (contentType) fastifyRes.header('content-type', contentType)
-      fastifyRes.code(401).send(errorResponse(invalidKeyError))
-      return
+    let retVal
+    if (auth && auth instanceof Function) {
+      try {
+        retVal = auth(key)
+      } catch (err) {
+        retVal = Promise.reject(err)
+      }
+    } else {
+      retVal = authenticate(keys, key)
     }
 
-    next()
+    const invalidKeyError = Error('invalid authorization header')
+    Promise.resolve(retVal).then((val) => {
+      if (!val) {
+        fastifyReq.log.error('invalid authorization header: `%s`', header)
+        if (contentType) fastifyRes.header('content-type', contentType)
+        fastifyRes.code(401).send(errorResponse(invalidKeyError))
+        return
+      }
+      next()
+    }).catch((err) => {
+      if (failSilent) {
+        fastifyRes.code(401).send(errorResponse(invalidKeyError))
+      } else {
+        fastifyRes.code(500).send(errorResponse(err instanceof Error ? err : Error(String(err))))
+      }
+    })
```

Comment:
I don’t think this whole promise block is needed. Or maybe I don’t understand its purpose.